File: src/batch-context.ts
import type { PhotoType } from "./analysis";

export type ContextAvailability = "available" | "partial" | "unavailable";
export type BatchGroupKind = "scene" | "lighting";
export type SequenceKind = "burst" | "exposure_bracket" | "repeated_frames";

export type ConsistencyNoteCode =
  | "exposure_reference"
  | "near_exposure_median"
  | "darker_than_group"
  | "brighter_than_group"
  | "near_white_balance_median"
  | "warmer_than_group"
  | "cooler_than_group"
  | "greener_than_group"
  | "more_magenta_than_group"
  | "bracket_member"
  | "partial_evidence"
  | "analysis_unavailable";

export interface ConsistencyNote {
  code: ConsistencyNoteCode;
  message: string;
}

export interface ExposureRelationship {
  delta_ev: number;
  confidence: number;
}

export interface WhiteBalanceRelationship {
  warm_cool_delta: number;
  green_magenta_delta: number;
  confidence: number;
}

export interface BatchGroup {
  group_id: string;
  asset_ids: string[];
  confidence: number;
  reference_candidate_ids: string[];
}

export interface SequenceGroup {
  group_id: string;
  asset_ids: string[];
  kind: SequenceKind;
  confidence: number;
  source_culling_group_id: string | null;
}

export interface ReferenceCandidate {
  group_kind: BatchGroupKind;
  group_id: string;
  asset_id: string;
  rank: number;
  technical_score: number;
  confidence: number;
  reasons: string[];
}

export interface AssetBatchContext {
  asset_id: string;
  availability: ContextAvailability;
  scene_group_id: string | null;
  lighting_group_id: string | null;
  sequence_group_id: string | null;
  reference_asset_id: string | null;
  exposure_delta_from_group: ExposureRelationship | null;
  wb_delta_from_group: WhiteBalanceRelationship | null;
  group_confidence: number;
  consistency_notes: ConsistencyNote[];
}

export interface BatchStageTimings {
  loading_ms: number;
  candidate_generation_ms: number;
  grouping_ms: number;
  context_ms: number;
  persistence_ms: number;
  total_ms: number;
}

export interface BatchDiagnostics {
  available_assets: number;
  partial_assets: number;
  unavailable_assets: number;
  candidate_comparisons: number;
  candidate_limit_per_asset: number;
  timings: BatchStageTimings;
  warnings: string[];
}

export interface BatchContext {
  schema_version: number;
  batch_id: string;
  job_id: string;
  photo_type: PhotoType;
  selected_asset_ids: string[];
  selection_identity: string;
  created_at: string;
  analysis_version: string;
  grouping_version: string;
  scene_groups: BatchGroup[];
  lighting_groups: BatchGroup[];
  sequence_groups: SequenceGroup[];
  asset_contexts: AssetBatchContext[];
  reference_candidates: ReferenceCandidate[];
  diagnostics: BatchDiagnostics;
}

export interface BatchContextRequest {
  job_id: string;
  photo_type: PhotoType;
  request_id: string;
  force: boolean;
}

export interface BatchContextProgress {
  job_id: string;
  request_id: string;
  photo_type: PhotoType;
  status: string;
  stage: string;
  completed: number;
  total: number;
  cached: boolean;
  duration_ms: number;
  error: string | null;
}

export interface BatchContextState {
  selected_count: number;
  selection_identity: string | null;
  context: BatchContext | null;
  cached: boolean;
  stale: boolean;
  progress: BatchContextProgress | null;
}

File: crates/photo-contracts/tests/batch_context.rs
use photo_contracts::{analysis::PhotoType, batch_context::*};

fn fixture() -> BatchContext {
    let scene = "1".repeat(64);
    let lighting = "2".repeat(64);
    BatchContext {
        schema_version: BATCH_CONTEXT_SCHEMA_VERSION,
        batch_id: "a".repeat(64),
        job_id: "job-1".into(),
        photo_type: PhotoType::Portrait,
        selected_asset_ids: vec!["asset-1".into()],
        selection_identity: "b".repeat(64),
        created_at: "2026-09-05T12:00:00Z".into(),
        analysis_version: "photo-analysis-schema-1".into(),
        grouping_version: "test-grouping-v1".into(),
        scene_groups: vec![BatchGroup {
            group_id: scene.clone(),
            asset_ids: vec!["asset-1".into()],
            confidence: 0.8,
            reference_candidate_ids: vec!["asset-1".into()],
        }],
        lighting_groups: vec![BatchGroup {
            group_id: lighting.clone(),
            asset_ids: vec!["asset-1".into()],
            confidence: 0.8,
            reference_candidate_ids: vec!["asset-1".into()],
        }],
        sequence_groups: vec![],
        asset_contexts: vec![AssetBatchContext {
            asset_id: "asset-1".into(),
            availability: ContextAvailability::Available,
            scene_group_id: Some(scene.clone()),
            lighting_group_id: Some(lighting.clone()),
            sequence_group_id: None,
            reference_asset_id: Some("asset-1".into()),
            exposure_delta_from_group: Some(ExposureRelationship {
                delta_ev: 0.,
                confidence: 0.8,
            }),
            wb_delta_from_group: Some(WhiteBalanceRelationship {
                warm_cool_delta: 0.,
                green_magenta_delta: 0.,
                confidence: 0.8,
            }),
            group_confidence: 0.8,
            consistency_notes: vec![ConsistencyNote {
                code: ConsistencyNoteCode::ExposureReference,
                message: "Reference".into(),
            }],
        }],
        reference_candidates: vec![
            ReferenceCandidate {
                group_kind: BatchGroupKind::Scene,
                group_id: scene,
                asset_id: "asset-1".into(),
                rank: 1,
                technical_score: 90.,
                confidence: 0.8,
                reasons: vec!["Stable source".into()],
            },
            ReferenceCandidate {
                group_kind: BatchGroupKind::Lighting,
                group_id: lighting,
                asset_id: "asset-1".into(),
                rank: 1,
                technical_score: 90.,
                confidence: 0.8,
                reasons: vec!["Stable source".into()],
            },
        ],
        diagnostics: BatchDiagnostics {
            available_assets: 1,
            partial_assets: 0,
            unavailable_assets: 0,
            candidate_comparisons: 0,
            candidate_limit_per_asset: 64,
            timings: BatchStageTimings::default(),
            warnings: vec![],
        },
    }
}

#[test]
fn versioned_contract_roundtrips_canonically() {
    let context = fixture();
    let json = context.canonical_json().unwrap();
    assert_eq!(BatchContext::parse(&json).unwrap(), context);
    assert!(!json.contains("recipe"));
    assert!(!json.contains("preset"));
}

#[test]
fn unknown_future_and_invalid_relationships_are_rejected() {
    let mut value = serde_json::to_value(fixture()).unwrap();
    value["schema_version"] = 2.into();
    assert_eq!(
        BatchContext::parse(&value.to_string()).unwrap_err(),
        BatchContextError::UnsupportedVersion(2)
    );
    let mut value = serde_json::to_value(fixture()).unwrap();
    value["edit_recipe"] = serde_json::json!({});
    assert!(BatchContext::parse(&value.to_string()).is_err());
    let mut context = fixture();
    context.asset_contexts[0]
        .exposure_delta_from_group
        .as_mut()
        .unwrap()
        .delta_ev = f64::NAN;
    assert!(context.validate().is_err());
    let mut context = fixture();
    context.selected_asset_ids.push("asset-1".into());
    assert!(context.validate().is_err());
}
